src/interpreter/Node.h

```diff
@@ -401,6 +389,25 @@ class ConditionalOperation {
     Own<Node> cond;
 };
 
+/**
+ * @class RelationalOperation
+ * @brief Interpreter operation that holds a single relation
+ */
+class RelationalOperation {
+public:
+    using RelationHandle = Own<RelationWrapper>;
+    RelationalOperation(RelationHandle* relHandle) : relHandle(relHandle) {}
```

Comment:
This is very strange to me.  A pointer to `Own<>`?  Is the handle optional?  If that's the case, I think it would be less confusing to just take `RelationWrapper*` instead of `Own<RelationWrapper>*`.  If it can't be null, use a reference, please.

Reply:
It uses 'pointer to `Own<>`' because the `Own<RelationWrapper>`  is actually managed by the engine.
Nodes do not manage the relation directly, `RelationHandle` is just a cache for fast lookup.

The design decision is made by the following reasoning:
1. The engine manages the runtime relation, let's say they stored in a `std::vector<Own<RelationWrapper>>`
2. For performance reason, we want the location of the target relation of each operation to be encoded statically during code generation in order to minimize the lookup time. (Let's say we use a `size_t index` in each `Node` to represent the index position of the target relation in the runtime environment, so far so good)
3. We have a special swap operation, whose semantic is to swap the content of two relations. For performance reason, we want to use `std::swap` instead of copying the whole content.
4. However, the above strategy raises an issue, once we "swap" the relation content, some previous statically encoded `size_t index`  can be invalided, as they are now pointing to the wrong target.
5. As a result, we use the current design.

Reply:
Also, currently, `RelationHandle` can still be null, but the reasons are more complicated. Basically, we have a `View` class that can provide faster lookup with a cached hint, but is not safe for parallelization and requires cost to create during runtime.

So for example, `Scan` does not require a `View` because it iterates through the whole relation and nothing needed to be cached, but `IndexScan` can benefit from a `View`.  Whatsmore, `ParallelIndexScan` again, can not use a `View` because it is not thread-safe.
However, intuitively, `ParallelIndexScan` would be a child of `IndexScan`, who is a child of `Scan`.

I hope those make sense to you, and I agree we need to redesign the class hierarchy in interpreter nodes. I'll set up a plan for this.

Reply:
Hi @XiaowenHu96,

Thank you for the response, it explains things well.

I still think that eventually not using `Own<RelationWrapper>*` is cleaner/less confusing because IMO it's not really idiomatic.  It also adds another "uncertainty point" to the code, because the "invariant" you describe is not encoded at the type-system level.  Specifically, now `Own<RelationWrapper>*` can be `nullptr` and perhaps `Own<RelationWrapper>` could also be `nullptr`.  From your explanation I think I understand that the latter is not possible but a new reader with no context may not know that.  Passing just `RelationWrapper*` does give me that information.

Perhaps you could adjust the constructor/and interface a little bit to make this clearer?  I will make a separate code suggestion.

Also, if my definition of 'less confusing' above doesn't convince you (which is fine!), perhaps the argument that it's more consistent with the rest of the codebase will.  For example, the child nodes of the ast are also held as `Own<[some child node]>`.  However, when calling `getChildNodes`, you get back a `std::vector<Node*>`, not `std::vector<Own<Node>*>`.  The codebase is quite consistent in this regard - `Own<X>` expresses ownership; `X*` means "someone else owns this.  So, to someone not familiar with the code base it's a nice consistency to have.  I think here you're introducing a "third" version, `Own<X>*`.  I hope that explains it.

In either case, thanks for doing this cleanup and taking the time to explain the internals.  It's useful as I'm learning more about the compiler.

Thanks,

Tom

Reply:
Hi @TomasPuverle 
I looked again at the code base and realize I made a mistake explaining it. It has been a while since I touched the interpreter code.

Basically, the current design is as follow:
1. engine: Owns the relations. Manage them in a `std::vector<Own<Own<RelationWrapper>>`, lets call the vector `Vec`.
2. node: Holds a pointer to `<Own<RelationWrapper>>` 
3. when `swap(i, j)`: engine swap the pointers in Vec[i] and Vec[j], i.e. `std::swap(Vec[i], Vec[j])`.

And in my first reply, I said a `size_t` index can be invalided by swap --- I was wrong. 
If we have engine holds `std::vector<Own<RelationWrapper>>`, and node holds an index, the swap should still be functional and the target of each node is still valid.

I now remember the real reason for choosing a pointer instead of just an index. Basically, I didn't want each operation to go through the engine, access the vector and grab the content. I want even faster access to the underlying relation by just goes through a pointer.

It seems like a trivial gain, but some innermost operations in Souffle can be executed more than millions of times. 
Unfortunately, I couldn't find the performance data for this, and I'm not even sure right now how much performance gain am I talking about -- It is a decision made a long time ago...

Let's say I want to still keep my mechanism because at the moment I don't have time to try others and benchmark them seriously. I think it is inevitable to either hold a pointer to `Own<RelationWrapper>` or a reference to `Own<RelationWrapper>`. I understand your concern about the first choice, and I agree with it completely. 
For the latter one, using a reference requires to redesign the class hierarchy. As I explained in my second reply, we currently having following inheritance:
`ParallelIndexScan` (No view, use RelationHandle) `<-` `IndexScan` (use View, no RelationHandle) `<-` Scan (No view, use RelationHandle)
As a result, `IndexScan` will hold a nullptr to the `Own<RelationWrapper>` and `ParallelIndexScan` will hold an invalid  `view`. I think the disaster here is because the nodes in the interpreter level do not related to each other in the same way as they were in the RAM level. Two operations can be very similar (IndexScan v.s. Scan) but have very different behaviour during execution and should not be made related.

I used to have only a single class for all the nodes in the interpreter, but people are not a fan of this. And I think the current class design is also not good either. I think the fix for this issue is to redesign the classes so that we can make the `relationHandle*` non-optional? That requires us to know whether an operation can use a cache ahead-of-time (at the time we design the class for that operation) and loses the ability to decide it during runtime, but we don't have this issue currently for our existed operations. 
Another fix I can think of is to wrap `Own<RelationWrapper>*` as a separated object which is then provided by the engine.
```
class RelationCache {
  Own<RelationWrapper>* cache;
public:
  std::optional<RelationWrapper> getRelation();
};
```
I would love to hear your idea.

Thanks again for providing suggestions and it is really helpful.

Regards,
Xiaowen

Reply:
By the way here is my email 
xihu5895@uni.sydney.edu.au
and my Skype:
xihu0208@gmail.com
In case you want to have a quick discussion with me :-)

Reply:
Hi @XiaowenHu96,

Thank you for the explanation and the offer to discuss!  I think your explanation makes sense and I like your thoughts and direction.  As such, it's probably not worth dwelling on this specific change since my observations are more stylistic at this point, and it sounds like you have additional improvements coming soon anyway.
I'm sure we'll get a chance to talk at some point in the future. :)

Thanks,

Tom